File: 01 Arrays and Strings/1.8-Zero-Matrix/ZeroMatrix.java
import java.util.*;
import java.io.*;

public class ZeroMatrix {

    public void zeroRow(int[][] M, int i) {
	for (int j = 0; j < M[i].length; ++j) {
	    M[i][j] = 0;
	}
    }

    public void zeroCol(int[][] M, int j) {
	for (int i = 0; i < M.length; ++i) {
	    M[i][j] = 0;
	}
    }

    public void zeroMatrix(int[][] M) {
	/* Keep track of which rows and columns have zeroes */
        boolean[] zrows = new boolean[M.length];
        boolean[] zcols = new boolean[M[0].length];

	for (int i = 0; i < M.length; ++i) {
	    for (int j = 0; j < M[0].length; ++j) {
		if (M[i][j] == 0) {
		    zrows[i] = true;
		    zcols[j] = true;
		}
	    }
	}

	for (int i = 0; i < zrows.length; ++i) {
	    if (zrows[i]) zeroRow(M, i);
	}

	for (int i = 0; i < zcols.length; ++i) {
	    if (zcols[i]) zeroCol(M, i);
	}
    }

    public static void printMatrix(int[][] M) {
	for (int[] v : M) {
	    for (int x : v) {
		System.out.print(x + " ");
	    }
	    System.out.println();
	}
    }

    public static int[][] readMatrix(InputStream stream) {
	Scanner in = new Scanner(stream);
	ArrayList<ArrayList<Integer>> rows = new ArrayList<ArrayList<Integer>>();
	while (in.hasNext()) {
	    ArrayList<Integer> row = new ArrayList<Integer>();
	    String line = in.nextLine();
	    for (String s : line.split("\\s+")) {
		row.add(Integer.parseInt(s));
	    }
	    rows.add(row);
	}

	int[][] M = new int[rows.size()][];
	int i = 0;
	for (ArrayList<Integer> row : rows) {
	    int[] data = new int[row.size()];
	    for (int j = 0; j < data.length; ++j) {
		data[j] = row.get(j);
	    }
	    M[i++] = data;
	}

	return M;
    }

    public static void main(String[] args) {
	System.out.println("Enter a matrix row by row (send EOF to end input): ");
	int[][] M = readMatrix(System.in);
	System.out.println("\nZero matrix:");
        (new ZeroMatrix()).zeroMatrix(M);
	printMatrix(M);
    }
}

File: 01_arrays_and_strings/cpp/zero_matrix.cpp
#include <iostream>
#include <vector>
#include <sstream>

void zeroRow(std::vector<std::vector<int>>& M, int i) {
  for (int j = 0; j < M[i].size(); ++j) {
    M[i][j] = 0;
  }
}

void zeroCol(std::vector<std::vector<int>>& M, int j) {
  for (int i = 0; i < M.size(); ++i) {
    M[i][j] = 0;
  }
}

void zeroMatrix(std::vector<std::vector<int>>& M) {
  /* Keep track of which rows and columns have zeroes */
  std::vector<bool> zrows(M.size());
  std::vector<bool> zcols(M[0].size());

  for (int i = 0; i < M.size(); ++i) {
    for (int j = 0; j < M[0].size(); ++j) {
      if (M[i][j] == 0) {
        zrows[i] = true;
	zcols[j] = true;
      }
    }
  }

  for (int i = 0; i < zrows.size(); ++i) {
    if (zrows[i]) zeroRow(M, i);
  }

    for (int i = 0; i < zcols.size(); ++i) {
      if (zcols[i]) zeroCol(M, i);
  }
}

void printMatrix(std::vector<std::vector<int>>& M) {
  for (std::vector<int> v : M) {
    for (int x : v) {
      std::cout << x << " ";
    }
    std::cout << std::endl;
  }
}

std::vector<std::vector<int>> readMatrix(std::istream& in) {
  std::vector<std::vector<int>> M;
  std::string line;
  while (getline(in, line)) {
    std::stringstream ls(line);
    int x;
    std::vector<int> v;
    while (ls >> x) {
      v.push_back(x);
    }
    M.push_back(v);
  }
  return M;
}

int main() {
  std::cout << "Enter a matrix row by row (send EOF to end input): " << std::endl;
  std::vector<std::vector<int>> M = readMatrix(std::cin);
  std::cout << std::endl << "Zero Matrix: " << std::endl;
  zeroMatrix(M);
  printMatrix(M);
  return 0;
}


